src/tasks/MotionForceTask.cpp

```diff
@@ -221,39 +227,55 @@ void MotionForceTask::updateTaskModel(const MatrixXd& N_prec) {
 		getConstRobotModel()->J(_link_name, _compliant_frame.translation());
 	_projected_jacobian = _jacobian * _N_prec;
 
-	MatrixXd range_pos =
-		Sai2Model::matrixRangeBasis(_projected_jacobian.topRows(3));
-	MatrixXd range_ori =
-		Sai2Model::matrixRangeBasis(_projected_jacobian.bottomRows(3));
-
-	_pos_range = range_pos.norm() == 0 ? 0 : range_pos.cols();
-	_ori_range = range_ori.norm() == 0 ? 0 : range_ori.cols();
-
-	if (_pos_range + _ori_range == 0) {
-		// there is no controllable degree of freedom for the task, just return
-		// should maybe print a warning here
-		_N.setIdentity(robot_dof, robot_dof);
-		return;
-	}
-
-	_current_task_range.setZero(6, _pos_range + _ori_range);
-	if (_pos_range > 0) {
-		_current_task_range.block(0, 0, 3, _pos_range) = range_pos;
-	}
-	if (_ori_range > 0) {
-		_current_task_range.block(3, _pos_range, 3, _ori_range) = range_ori;
-	}
-
-	Sai2Model::OpSpaceMatrices op_space_matrices =
-		getConstRobotModel()->operationalSpaceMatrices(
-			_current_task_range.transpose() * _projected_jacobian);
-	_Lambda = op_space_matrices.Lambda;
-	_Jbar = op_space_matrices.Jbar;
-	_N = op_space_matrices.N;
+	// /*
+	// 	Original implementation 
+	// */
+	// MatrixXd range_pos =
+	// 	Sai2Model::matrixRangeBasis(_projected_jacobian.topRows(3), 1e-6);
+	// MatrixXd range_ori =
+	// 	Sai2Model::matrixRangeBasis(_projected_jacobian.bottomRows(3), 1e-6);
+
+	// _pos_range = range_pos.norm() == 0 ? 0 : range_pos.cols();
+	// _ori_range = range_ori.norm() == 0 ? 0 : range_ori.cols();
+
+	// if (_pos_range + _ori_range == 0) {
+	// 	// there is no controllable degree of freedom for the task, just return
+	// 	// should maybe print a warning here
+	// 	_N.setIdentity(robot_dof, robot_dof);
+	// 	return;
+	// }
+
+	// _current_task_range.setZero(6, _pos_range + _ori_range);
+	// if (_pos_range > 0) {
+	// 	_current_task_range.block(0, 0, 3, _pos_range) = range_pos;
+	// }
+	// if (_ori_range > 0) {
+	// 	_current_task_range.block(3, _pos_range, 3, _ori_range) = range_ori;
+	// }
+
+	// Sai2Model::OpSpaceMatrices op_space_matrices =
+	// 	getConstRobotModel()->operationalSpaceMatrices(
+	// 		_current_task_range.transpose() * _projected_jacobian);
+	// _Lambda = op_space_matrices.Lambda;
+	// _Jbar = op_space_matrices.Jbar;
+	// _N = op_space_matrices.N;
+
+	/*
+		Singularity handler implementation 
+	*/
+	SingularityOpSpaceMatrices singularity_op_space_matrices = _singularity_handler->updateTaskModel(_projected_jacobian, _N_prec);
+	_projected_jacobian_ns = singularity_op_space_matrices.projected_jacobian_ns;
+	_Lambda_ns = singularity_op_space_matrices.Lambda_ns;
+	_task_range_ns = singularity_op_space_matrices.task_range_ns;
+	_projected_jacobian_s = singularity_op_space_matrices.projected_jacobian_s;
+	_Lambda_s = singularity_op_space_matrices.Lambda_s;
+	_task_range_s = singularity_op_space_matrices.task_range_s;
 
 	switch (_dynamic_decoupling_type) {
```

Comment:
The more I think about it, the more I think we should remove the partial dynamic decoupling. Please do it if you move the decoupling computations inside the new class as I suggested before